strings/boyer_moore.cpp

```diff
@@ -0,0 +1,270 @@
+ *
+ * @param str Text we are seatching in.
+ * @param arg pattern structure containing the preprocessed pattern
+ * @return (vector of) indexes of the occurrences of pattern in text
+ */
+std::vector<size_t> search(const std::string& str, const pattern& arg) {
+    size_t index_position = arg.pat.size() - 1;
+    std::vector<size_t> index_storage;
+
+    while (index_position < str.length()) {
+        size_t index_string = index_position;
+        int index_pattern = static_cast<int>(arg.pat.size()) - 1;
+
+        while (index_pattern >= 0 &&
+               str[index_string] == arg.pat[index_pattern]) {
+            --index_pattern;
+            --index_string;
+        }
+
+        if (index_pattern < 0) {
+            index_storage.push_back(index_position - arg.pat.length() + 1);
+            index_position += arg.good_suffix[0];
+        } else {
+            index_position += std::max(arg.bad_char[str[index_string]],
+                                       arg.good_suffix[index_pattern + 1]);
+        }
+    }
+
+    return index_storage;
+}
+
+/**
+ * @brief Check if pat is prefix of str.
+ *
+ * @param str pointer to some part of the input text.
+ * @param pat the searched pattern.
+ * @param len length of the searched pattern
+ * @returns `true` if pat IS prefix of str.
+ * @returns `false` if pat is NOT a prefix of str.
+ */
+bool is_prefix(const char* str, const char* pat, size_t len) {
+    if (strlen(str) < len) {
+        return false;
+    }
+
+    for (size_t i = 0; i < len; i++) {
+        if (str[i] != pat[i]) {
+            return false;
+        }
+    }
+
+    return true;
+}
+}  // namespace boyer_moore
+}  // namespace strings
+/**
+ * @brief A test case in which we search for every appearance of the word 'and'
+ * @param text The text in which we search for appearance of the word 'and'
+ * @returns void
+ */
+void and_test(const char* text) {
+    strings::boyer_moore::pattern ands;
+    strings::boyer_moore::init_pattern("and", ands);
+    std::vector<size_t> indexes = strings::boyer_moore::search(text, ands);
+
+    assert(indexes.size() == 2);
+    assert(strings::boyer_moore::is_prefix(text + indexes[0], "and", 3));
+    assert(strings::boyer_moore::is_prefix(text + indexes[1], "and", 3));
+}
+
+/**
+ * @brief  A test case in which we search for every appearance of the word 'pat'
+ * @param text The text in which we search for appearance of the word 'pat'
+ * @returns void
+ */
+void pat_test(const char* text) {
+    strings::boyer_moore::pattern pat;
+    strings::boyer_moore::init_pattern("pat", pat);
+    std::vector<size_t> indexes = strings::boyer_moore::search(text, pat);
+
+    assert(indexes.size() == 6);
+
+    for (const auto& idx : indexes) {
+        assert(strings::boyer_moore::is_prefix(text + idx, "pat", 3));
+    }
+}
+/**
+ * @brief Self-test implementations
+ * @returns void
+ */
+static void tests() {
+    const char* text =
+        "When pat Mr. and Mrs. pat Dursley woke up on the dull, gray  \
```

Comment:
Nitpick Shouldn't it be Mr. pat and Mrs pat

Reply:
This is text I copy-pasted from random book (Harry Potter I think).

I added some "pat" substrings so I could test the search. The text isn't supposed to make any sense 😄

Reply:
Ah no wonder! thank you for the clarification!